test/MiniMeToken.t.sol

```diff
@@ -115,16 +427,253 @@ contract AllowanceTest is MiniMeTokenTest {
+    }
+    /**
+     * @dev decodes sig of abi encoded call
+     * @param _data abi encoded data
+     * @return sig (first 4 bytes)
+     */
+
+    function abiDecodeSig(bytes memory _data) private pure returns (bytes4 sig) {
+        assembly {
+            sig := mload(add(_data, add(0x20, 0)))
+        }
+    }
+
+    /**
+     * @dev get a slice of byte array
+     * @param _bytes source
+     * @param _start pointer
+     * @param _length size to read
+     * @return sliced bytes
+     */
+    function slice(bytes memory _bytes, uint256 _start, uint256 _length) private pure returns (bytes memory) {
+        require(_bytes.length >= (_start + _length));
+
+        bytes memory tempBytes;
+
+        assembly {
+            switch iszero(_length)
+            case 0 {
+                // Get a location of some free memory and store it in tempBytes as
+                // Solidity does for memory variables.
+                tempBytes := mload(0x40)
+
+                // The first word of the slice result is potentially a partial
+                // word read from the original array. To read it, we calculate
+                // the length of that partial word and start copying that many
+                // bytes into the array. The first word we copy will start with
+                // data we don't care about, but the last `lengthmod` bytes will
+                // land at the beginning of the contents of the new array. When
+                // we're done copying, we overwrite the full first word with
+                // the actual length of the slice.
+                let lengthmod := and(_length, 31)
+
+                // The multiplication in the next line is necessary
+                // because when slicing multiples of 32 bytes (lengthmod == 0)
+                // the following copy loop was copying the origin's length
+                // and then ending prematurely not copying everything it should.
+                let mc := add(add(tempBytes, lengthmod), mul(0x20, iszero(lengthmod)))
+                let end := add(mc, _length)
+
+                for {
+                    // The multiplication in the next line has the same exact purpose
+                    // as the one above.
+                    let cc := add(add(add(_bytes, lengthmod), mul(0x20, iszero(lengthmod))), _start)
+                } lt(mc, end) {
+                    mc := add(mc, 0x20)
+                    cc := add(cc, 0x20)
+                } { mstore(mc, mload(cc)) }
+
+                mstore(tempBytes, _length)
+
+                //update free-memory pointer
+                //allocating the array padded to 32 bytes like the compiler does now
+                mstore(0x40, and(add(mc, 31), not(31)))
+            }
+            //if we want a zero-length slice let's just return a zero-length array
+            default {
+                tempBytes := mload(0x40)
+
+                mstore(0x40, add(tempBytes, 0x20))
+            }
+        }
+
+        return tempBytes;
```

Comment:
We migtht want to consider moving these contracts used for tests out of the test file as they bloat the test file and they contain quite a bit of logic in them.